sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala

```diff
@@ -225,10 +227,20 @@ final class SpecificMutableRow(val values: Array[MutableValue])
     val newValues = new Array[Any](values.length)
     var i = 0
     while (i < values.length) {
-      newValues(i) = values(i).boxed
+      val value = values(i).boxed
+      if (value.isInstanceOf[UTF8String]) {
+        newValues(i) = value.asInstanceOf[UTF8String].clone()
+      } else if (value.isInstanceOf[InternalRow]) {
+        newValues(i) = value.asInstanceOf[InternalRow].copy()
+      } else if (value.isInstanceOf[ArrayData]) {
+        newValues(i) = value.asInstanceOf[ArrayData].copy()
+      } else if (value.isInstanceOf[MapData]) {
+        newValues(i) = value.asInstanceOf[MapData].copy()
+      } else {
+        newValues(i) = value
```

Comment:
this code is duplicated 4 times, should we introduce a trait for it? (I'm not able to create a util function for it because these codes differ a little)